kanidmd/src/lib/value.rs

```diff
@@ -281,6 +288,7 @@ pub enum PartialValue {
     Cid(Cid),
     Nsuniqueid(String),
     DateTime(OffsetDateTime),
+    EmailAddress(String),
```

Comment:
How come everything is a `String`? Do we not have more explicit types for things? Why not something like:

```suggestion
    EmailAddress {
        local: String,
        domain: String,
    },
```

Reply:
Because I am not always aware if there are better types. Value.rs does need to be rewritter because I didn't know how to make borrow as a trait work in the past between value and partialvalue. But because value is fully in memory, we can and freely change these types if we find something better in the future.

Reply:
What's the functional benefit of the local/domain breakdown for an email address enum? They're essentially arbitrary strings with some validation - like usernames/phone numbers/physical addresses.

Reply:
That's a good point too yaleman. We do validate the content as it comes in so we know the string data is correct too. Saying this, database corruption could be a thing, but even making this a stricter type won't protect us from that ...

Reply:
My thinking was that although all email addresses are `String`s, the opposite isn't necessarily true. This leaves anyone looking at the code to assume that the `String` field is a valid email address, which I wasn't certain of until William's comment. Also, it's odd that you can unpack the `EmailAddress` variant and get a `String`, and then it can do anything that a `String` can do that an email address shouldn't be able to do, like join it with another `String`.

But if we had an explicit `EmailAddr` type, we could give it special email address functionality (not sure exactly what this would look like though) and make it generally more robust. Thus, the variant would look like:

```suggestion
    EmailAddress(EmailAddr)
```

Where `EmailAddr` is declared as
```rust
struct EmailAddr {
    local: String,
    domain: String,
}
```

This could probably be used in many other places as well. But if we tread carefully and don't use this in a lot of places, you're both right that it won't make much difference at all.

Reply:
Well, remember this is also an inmemory representation. So we can change it to local/domain when and if we find a need for that. Today, we don't have that need, so what benefit does it bring? Remember a key part of this design is that we split the DB (stable, can't change format) from the in memory (flexible, can change, adapt over time), so if we find out tomorrow we need this split, we can doit immediately :) 

So for now, I think unless we have a need to do this, we should leave it. And when we do have the need for it, we can easily retro fit this in.